pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/steps/IdentifierGenerator.scala

```diff
@@ -1,41 +1,48 @@
 package uk.ac.wellcome.platform.idminter.steps
 
+import grizzled.slf4j.Logging
 import uk.ac.wellcome.models.work.internal.SourceIdentifier
 import uk.ac.wellcome.platform.idminter.database.IdentifiersDao
 import uk.ac.wellcome.platform.idminter.models.Identifier
 import uk.ac.wellcome.platform.idminter.utils.Identifiable
 
-import scala.util.Try
+import scala.collection.breakOut
+import scala.util.{Success, Try}
 
-class IdentifierGenerator(identifiersDao: IdentifiersDao) {
-
-  def retrieveOrGenerateCanonicalId(
-    identifier: SourceIdentifier
-  ): Try[String] =
-    Try {
-      identifiersDao
-        .lookupId(
-          sourceIdentifier = identifier
-        )
-        .flatMap {
-          case Some(id) => Try(id.CanonicalId)
-          case None     => generateAndSaveCanonicalId(identifier)
-        }
-    }.flatten
-
-  private def generateAndSaveCanonicalId(
-    sourceIdentifier: SourceIdentifier
-  ): Try[String] = {
-    val canonicalId = Identifiable.generate
+class IdentifierGenerator(identifiersDao: IdentifiersDao) extends Logging {
+  // This function fetches canonicalIds for sourceIdentifiers, and generates
+  // and saves canonicalIds where it can't find existing ones. For this reason
+  // it's important not to run it in situations where there might be a race
+  // condition (IDs being inserted in between lookup and update).
+  def retrieveOrGenerateCanonicalIds(sourceIdentifiers: Seq[SourceIdentifier])
+    : Try[Map[SourceIdentifier, Identifier]] =
     identifiersDao
-      .saveIdentifier(
-        Identifier(
-          canonicalId = canonicalId,
-          sourceIdentifier = sourceIdentifier
-        )
-      )
-      .map { _ =>
-        canonicalId
+      .lookupIds(sourceIdentifiers)
+      .flatMap {
+        case IdentifiersDao.LookupResult(foundMap, notFoundIdentifiers) =>
+          generateAndSaveCanonicalIds(notFoundIdentifiers).map {
+            newIdentifiers =>
+              val newMap: Map[SourceIdentifier, Identifier] =
+                (notFoundIdentifiers zip newIdentifiers)(breakOut)
+              foundMap ++ newMap
```

Comment:
See comment earlier; picking better names than "found" or "not found" would make the path to `newIdentifiers` clearer.